include/xgboost/json.h

```diff
@@ -0,0 +1,530 @@
+ *   // Assign key "arr" with a empty JSON Array;
+ *   object["arr"] = Array();
+ * \endcode
+ */
+class Json {
+  friend JsonWriter;
+
+ public:
+  /*! \brief Load a Json object from string. */
+  static Json Load(StringView str, bool ignore_specialization = false);
+  /*! \brief Pass your own JsonReader. */
+  static Json Load(JsonReader* reader);
+  /*! \brief Dump json into stream. */
+  static void Dump(Json json, std::ostream* stream,
+                   bool pretty = ConsoleLogger::ShouldLog(
+                       ConsoleLogger::LogVerbosity::kDebug));
+
+  Json() : ptr_{new JsonNull} {}
+
+  // number
+  explicit Json(JsonNumber number) : ptr_{new JsonNumber(number)} {}
+  Json& operator=(JsonNumber number) {
+    ptr_.reset(new JsonNumber(std::move(number)));
+    return *this;
+  }
+
+  // array
+  explicit Json(JsonArray list) :
+      ptr_ {new JsonArray(std::move(list))} {}
+  Json& operator=(JsonArray array) {
+    ptr_.reset(new JsonArray(std::move(array)));
+    return *this;
+  }
+
+  // raw
+  explicit Json(JsonRaw str) :
+      ptr_{new JsonRaw(std::move(str))} {}
+  Json& operator=(JsonRaw str) {
+    ptr_.reset(new JsonRaw(std::move(str)));
+    return *this;
+  }
+
+  // object
+  explicit Json(JsonObject object) :
+      ptr_{new JsonObject(std::move(object))} {}
+  Json& operator=(JsonObject object) {
+    ptr_.reset(new JsonObject(std::move(object)));
+    return *this;
+  }
+  // string
+  explicit Json(JsonString str) :
+      ptr_{new JsonString(std::move(str))} {}
+  Json& operator=(JsonString str) {
+    ptr_.reset(new JsonString(std::move(str)));
+    return *this;
+  }
+  // bool
+  explicit Json(JsonBoolean boolean) :
+      ptr_{new JsonBoolean(std::move(boolean))} {}
+  Json& operator=(JsonBoolean boolean) {
+    ptr_.reset(new JsonBoolean(std::move(boolean)));
+    return *this;
+  }
+  // null
+  explicit Json(JsonNull null) :
+      ptr_{new JsonNull(std::move(null))} {}
+  Json& operator=(JsonNull null) {
+    ptr_.reset(new JsonNull(std::move(null)));
+    return *this;
+  }
+
+  // copy
+  Json(Json const& other) : ptr_{other.ptr_} {}
+  Json& operator=(Json const& other);
+  // move
+  Json(Json&& other) : ptr_{std::move(other.ptr_)} {}
+  Json& operator=(Json&& other) {
+    ptr_ = std::move(other.ptr_);
+    return *this;
+  }
+
+  /*! \brief Index Json object with a std::string, used for Json Object. */
+  Json& operator[](std::string const & key) const { return (*ptr_)[key]; }
+  /*! \brief Index Json object with int, used for Json Array. */
+  Json& operator[](int ind)                 const { return (*ptr_)[ind]; }
+
+  /*! \Brief Return the reference to stored Json value. */
+  Value const& GetValue() const & { return *ptr_; }
+  Value const& GetValue() &&      { return *ptr_; }
+  Value&       GetValue() &       { return *ptr_; }
+
+  bool operator==(Json const& rhs) const {
+    return *ptr_ == *(rhs.ptr_);
+  }
+
+ private:
+  std::shared_ptr<Value> ptr_;
```

Comment:
What do you think about using std::unique_ptr here
I looked though this pull request and your another (Json format) pull request and it seems we never actually have two Json objects pointing to the same instance of Value. We always move them around.

Benefits of unique_ptr is
1. easier to reason
2. no cycles
3. shared ownership could lead to hard-to-debug problems

Reply:
I plan to implement a pool allocator in combination with dmlc:: Any in the future for better performance.